### PR TITLE
SB-14589: content review api

### PR DESCRIPTION
**Pull request by krgauraw**
<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/SB-14589" title="SB-14589" target="_blank">TaskSB-14589</a>  Content Review API refactoring
  </td></table>
  <br />
 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions: Java 11, scala-2.11, play-2.7.2
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/PluginMimeTypeMgrImpl.scala [hunk 1]
```diff
@@ -58,4 +59,8 @@ class PluginMimeTypeMgrImpl(implicit ss: StorageService) extends BaseMimeTypeMan
 			Map[String, AnyRef]("semanticVersion" -> version, "targets" -> targetList)
 		}else Map[String, AnyRef]()
 	}
+
+	override def review(objectId: String, node: Node)(implicit ec: ExecutionContext, ontologyEngineContext: OntologyEngineContext): Future[Map[String, AnyRef]] = {
+		Future(getEnrichedMetadata(node.getMetadata.getOrDefault("status", "").asInstanceOf[String]))
```
Comment (maheshkumargangula) [inline platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/PluginMimeTypeMgrImpl.scala:64]: For plugin we need to have the artifactUrl validation. Please check and enable it.

File: platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImpl.scala [hunk 2]
```diff
@@ -90,4 +95,30 @@ class EcmlMimeTypeMgrImpl(implicit ss: StorageService) extends BaseMimeTypeManag
 		}
 	}
 
+	override def review(objectId: String, node: Node)(implicit ec: ExecutionContext, ontologyEngineContext: OntologyEngineContext): Future[Map[String, AnyRef]] = {
+		validate(node)
+		Future(getEnrichedMetadata(node.getMetadata.getOrDefault("status", "").asInstanceOf[String]))
+	}
+
+	def validate(node: Node)(implicit ec: ExecutionContext, oec: OntologyEngineContext): Unit = {
+		val artifactUrl = node.getMetadata.getOrDefault("artifactUrl", "").asInstanceOf[String]
```
Comment (maheshkumargangula) [inline platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImpl.scala:104]: Why don't you write a method in Node class to get artifactUrl?

File: platform-modules/mimetype-manager/src/test/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImplTest.scala [hunk 3]
```diff
@@ -80,6 +82,29 @@ class EcmlMimeTypeMgrImplTest extends AsyncFlatSpec with Matchers with AsyncMock
...
..."type\":\"json\"},{\"id\":\"org.ekstep.text\",\"plugin\":\"org.ekstep.text\",\"ver\":\"1.2\",\"src\":\"/content-plugins/org.ekstep.text-1.2/renderer/supertextplugin.js\",\"type\":\"plugin\"},{\"id\":\"org.ekstep.text_manifest\",\"plugin\":\"org.ekstep.text\",\"ver\":\"1.2\",\"src\":\"/content-plugins/org.ekstep.text-1.2/manifest.json\",\"type\":\"json\"},{\"id\":\"do_2133122975753093121715\",\"src\":\"https://www.youtube.com/watch?v=YH7p6ncEb6g\",\"assetId\":\"do_2133122975753093121715\",\"type\":\"youtube\"},{\"id\":\"do_2133016737715240961616\",\"src\":\"/assets/public/content/do_2133016737715240961616/artifact/do_2133016737715240961616_1623739474410_pexels-photo-594364.jpeg\",\"type\":\"image\"},{\"id\":\"do_2133016742018826241623\",\"src\":\"/assets/public/content/do_2133016742018826241623/artifact/do_2133016742018826241623_1623739526722_.png\",\"type\":\"image\"}]},\"plugin-manifest\":{\"plugin\":[{\"id\":\"org.ekstep.navigation\",\"ver\":\"1.0\",\"type\":\"plugin\",\"depends\":\"\"},{\"id\":\"org.ekstep.video\",\"ver\":\"1.5\",\"type\":\"widget\",\"depends\":\"\"},{\"id\":\"org.ekstep.text\",\"ver\":\"1.2\",\"type\":\"plugin\",\"depends\":\"\"}]},\"compatibilityVersion\":4}}"
```
Comment (maheshkumargangula) [inline platform-modules/mimetype-manager/src/test/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImplTest.scala:89]: Why don't you use """ for having proper JSON.

File: platform-modules/mimetype-manager/src/test/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImplTest.scala [hunk 4]
```diff
@@ -80,6 +82,29 @@ class EcmlMimeTypeMgrImplTest extends AsyncFlatSpec with Matchers with AsyncMock
...
+        node.setObjectType("Content")
+        node.getMetadata.put("artifactUrl", "https://sunbirddev.blob.core.windows.net/sunbird-content-dev/content/do_113105564164997120111/artifact/1599796728064_do_11310551225702809612421.zip")
+        val identifier = "do_2133122972131737601714"
+        val resFuture = new EcmlMimeTypeMgrImpl().review(identifier, node)
+        resFuture.map(result => {
+            assert(null != result)
+            assert(result.nonEmpty)
+            assert("Review" == result.getOrElse("status",""))
+        })
+        assert(true)
```
Comment (maheshkumargangula) [inline platform-modules/mimetype-manager/src/test/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImplTest.scala:105]:
```suggestion
```

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

File: platform-modules/mimetype-manager/src/test/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImplTest.scala (hunk 4 above)
Comment (krgauraw) [reply platform-modules/mimetype-manager/src/test/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImplTest.scala]: done

File: platform-modules/mimetype-manager/src/test/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImplTest.scala (hunk 3 above)
Comment (krgauraw) [reply platform-modules/mimetype-manager/src/test/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImplTest.scala]: done

File: platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImpl.scala (hunk 2 above)
Comment (krgauraw) [reply platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/EcmlMimeTypeMgrImpl.scala:104]: done

File: platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/PluginMimeTypeMgrImpl.scala (hunk 1 above)
Comment (krgauraw) [reply platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/PluginMimeTypeMgrImpl.scala:66]: done

File: platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/CollectionMimeTypeMgrImpl.scala [hunk 5]
```diff
@@ -2,20 +2,88 @@ package org.sunbird.mimetype.mgr.impl
...
+			if (ResponseHandler.checkError(response) && !ResponseHandler.isResponseNotFoundError(response)) {
+				throw new ServerException("ERR_COLLECTION_REVIEW", "Unable to fetch hierarchy for identifier:" + rootNode.getIdentifier)
+			} else if (ResponseHandler.checkError(response) && ResponseHandler.isResponseNotFoundError(response)) {
+				throw new ClientException("ERR_COLLECTION_REVIEW", "No hierarchy is present in cassandra for identifier:" + rootNode.getIdentifier)
+			} else Future(response.getResult.toMap.getOrElse("hierarchy", "{}").asInstanceOf[String])
+		})
+	}
+
+	def getChildren(hierarchyStr: String): Map[String, AnyRef] = {
+		val hierarchy: Map[String, AnyRef] = if (!hierarchyStr.asInstanceOf[String].isEmpty) {
```
[the remark this answers is not in the document]
Comment (krgauraw) [reply platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/CollectionMimeTypeMgrImpl.scala]: done

File: platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/CollectionMimeTypeMgrImpl.scala [hunk 6]
```diff
@@ -2,20 +2,88 @@ package org.sunbird.mimetype.mgr.impl
...
+			})
+		}).flatMap(f => f)
+	}
+
+	def getCollectionHierarchy(request: Request, rootNode: Node)(implicit ec: ExecutionContext, oec: OntologyEngineContext): Future[Any] = {
+		oec.graphService.readExternalProps(request, List("hierarchy")).flatMap(response => {
+			if (ResponseHandler.checkError(response) && !ResponseHandler.isResponseNotFoundError(response)) {
+				throw new ServerException("ERR_COLLECTION_REVIEW", "Unable to fetch hierarchy for identifier:" + rootNode.getIdentifier)
+			} else if (ResponseHandler.checkError(response) && ResponseHandler.isResponseNotFoundError(response)) {
+				throw new ClientException("ERR_COLLECTION_REVIEW", "No hierarchy is present in cassandra for identifier:" + rootNode.getIdentifier)
```
[the remark this answers is not in the document]
Comment (krgauraw) [reply platform-modules/mimetype-manager/src/main/scala/org/sunbird/mimetype/mgr/impl/CollectionMimeTypeMgrImpl.scala]: done

File: content-api/content-actors/src/test/scala/org/sunbird/content/actors/TestContentActor.scala [hunk 7]
```diff
@@ -406,6 +406,67 @@ class TestContentActor extends BaseSpec with MockFactory {
...
+        node.getMetadata.put("contentType", "Resource")
+        node.getMetadata.put("status", "Processing")
+        node.getMetadata.put("organisationBoardIds", new util.ArrayList[String](){{add("ncf_board_cbse")}})
+        (graphDB.getNodeByUniqueId(_: String, _: String, _: Boolean, _: Request)).expects(*, *, *, *).returns(Future(node)).anyNumberOfTimes()
+        (graphDB.readExternalProps(_: Request, _: List[String])).expects(*, *).returns(Future(new Response())).anyNumberOfTimes()
+        val request = getContentRequest()
+        request.getContext.put("identifier", "do_123")
+        request.setOperation("reviewContent")
+        val response = callActor(request, Props(new ContentActor()))
+        assert(response.getResponseCode == ResponseCode.CLIENT_ERROR)
```
[the remark this answers is not in the document]
Comment (krgauraw) [reply content-api/content-actors/src/test/scala/org/sunbird/content/actors/TestContentActor.scala:446]: done

File: content-api/content-actors/src/main/scala/org/sunbird/content/review/mgr/ReviewManager.scala [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+import scala.collection.Map
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+
+object ReviewManager {
+
+	def review(request: Request, node: Node)(implicit oec: OntologyEngineContext, ec: ExecutionContext): Future[Response] = {
+		val identifier: String = node.getIdentifier
+		val mimeType = node.getMetadata().getOrDefault("mimeType", "assets").asInstanceOf[String]
```
[the remark this answers is not in the document]
Comment (krgauraw) [reply content-api/content-actors/src/main/scala/org/sunbird/content/review/mgr/ReviewManager.scala]: removed

File: content-api/content-actors/src/main/scala/org/sunbird/content/actors/ContentActor.scala [hunk 9]
```diff
@@ -146,6 +148,21 @@ class ContentActor @Inject() (implicit oec: OntologyEngineContext, ss: StorageSe
 
 	def importContent(request: Request): Future[Response] = importMgr.importObject(request)
 
+	def reviewContent(request: Request): Future[Response] = {
+		val identifier: String = request.getContext.getOrDefault("identifier", "").asInstanceOf[String]
+		val readReq = new Request(request)
+		readReq.put("identifier", identifier)
+		readReq.put("mode", "edit")
+		readReq.put("fields", new util.ArrayList[String])
```
[the remark this answers is not in the document]
Comment (krgauraw) [reply content-api/content-actors/src/main/scala/org/sunbird/content/actors/ContentActor.scala]: done. removed

**Pull request merged**